File: remote-claude/SKILLS/mcp-builder/examples/basic-server.ts
/**
 * Basic MCP server example with TypeScript SDK
 */

import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { z } from "zod";

// Constants
const CHARACTER_LIMIT = 25000;
const API_BASE_URL = "https://api.example.com";

// Initialize server
const server = new Server(
  {
    name: "example-server",
    version: "1.0.0",
  },
  {
    capabilities: {
      tools: {},
    },
  }
);

// Input validation schema
const SearchInputSchema = z.object({
  query: z.string().min(1).max(500),
  limit: z.number().int().min(1).max(100).default(10),
  format: z.enum(["json", "markdown"]).default("json"),
}).strict();

// Register tool
server.registerTool({
  name: "search_items",
  description: "Search for items in the database",
  inputSchema: SearchInputSchema,
  annotations: {
    readOnlyHint: true,
    destructiveHint: false,
    idempotentHint: true,
    openWorldHint: true,
  },
}, async ({ query, limit = 10, format = "json" }): Promise<string> => {
  try {
    // Validate input
    const validated = SearchInputSchema.parse({ query, limit, format });

    // API call logic here
    const results = await fetchFromApi(validated.query, validated.limit);

    // Format response
    if (validated.format === "markdown") {
      return formatAsMarkdown(results);
    }
    return formatAsJson(results);

  } catch (error) {
    if (error instanceof z.ZodError) {
      return `Error: Invalid input - ${error.message}`;
    }
    return `Error: Search failed - ${error.message}`;
  }
});

// Server startup
async function main(): Promise<void> {
  const transport = new StdioServerTransport();
  await server.connect(transport);
}

main().catch((error) => {
  console.error("Server error:", error);
  process.exit(1);
});

File: remote-claude/SKILLS/mcp-builder/examples/basic-server.py
"""Basic MCP server example with Python SDK"""

from mcp.server import Server
from mcp.server.stdio import stdio_server
from pydantic import BaseModel, Field

# Constants
CHARACTER_LIMIT = 25000
API_BASE_URL = "https://api.example.com"

# Initialize server
mcp = Server("example-server")

# Input validation model
class SearchInput(BaseModel):
    query: str = Field(..., min_length=1, max_length=500, description="Search query")
    limit: int = Field(10, ge=1, le=100, description="Max results (1-100)")
    format: str = Field("json", pattern="^(json|markdown)$", description="Response format")

@mcp.tool()
async def search_items(query: str, limit: int = 10, format: str = "json") -> str:
    """
    Search for items in the database.

    Args:
        query: Search query string
        limit: Maximum number of results (1-100)
        format: Response format ('json' or 'markdown')

    Returns:
        Formatted search results (JSON or Markdown)

    Example:
        search_items(query="machine learning", limit=5, format="markdown")
    """
    try:
        # Validate input
        validated = SearchInput(query=query, limit=limit, format=format)

        # API call logic here
        results = await fetch_from_api(validated.query, validated.limit)

        # Format response
        if validated.format == "markdown":
            return format_as_markdown(results)
        return format_as_json(results)

    except ValueError as e:
        return f"Error: Invalid input - {str(e)}"
    except Exception as e:
        return f"Error: Search failed - {str(e)}"

# Server startup
async def main():
    async with stdio_server() as (read_stream, write_stream):
        await mcp.run(read_stream, write_stream, mcp.create_initialization_options())

if __name__ == "__main__":
    import asyncio
    asyncio.run(main())
